File: libs/tsevents/src/envelope.ts
import { randomUUID } from 'crypto';

export interface Actor { id: number; role: string }

export interface Envelope<T = any> {
  id: string;
  type: string;
  version: number;
  occurred_at: string;
  correlation_id: string;
  causation_id: string | null;
  actor: Actor | null;
  payload: T;
}

export function newEnvelope<T>(type: string, payload: T, opts: {
  correlationId?: string; causationId?: string; actor?: Actor; version?: number;
} = {}): Envelope<T> {
  return {
    id: randomUUID(),
    type,
    version: opts.version ?? 1,
    occurred_at: new Date().toISOString(),
    correlation_id: opts.correlationId ?? randomUUID(),
    causation_id: opts.causationId ?? null,
    actor: opts.actor ?? null,
    payload,
  };
}

/** Derive a follow-on event that keeps the saga linked. */
export function caused<T>(parent: Envelope, type: string, payload: T): Envelope<T> {
  return newEnvelope(type, payload, {
    correlationId: parent.correlation_id,
    causationId: parent.id,
    actor: parent.actor ?? undefined,
  });
}

File: libs/pyevents/pyevents/envelope.py
"""The message envelope shared by every service, in every language.

`correlation_id` groups every event belonging to one business transaction;
`causation_id` names the message that directly caused this one. Together they
let a whole saga be reconstructed from logs without distributed tracing.
"""
from __future__ import annotations
import json, uuid
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


@dataclass
class Envelope:
    type: str
    payload: dict
    version: int = 1
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    occurred_at: str = field(default_factory=_now)
    correlation_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    causation_id: str | None = None
    actor: dict | None = None

    def to_bytes(self) -> bytes:
        return json.dumps(asdict(self), separators=(",", ":")).encode()

    @staticmethod
    def from_bytes(raw: bytes) -> "Envelope":
        d = json.loads(raw.decode())
        return Envelope(**{k: d.get(k) for k in
                           ("type", "payload", "version", "id", "occurred_at",
                            "correlation_id", "causation_id", "actor")})

    def caused(self, type_: str, payload: dict) -> "Envelope":
        """Derive a follow-on event that keeps the saga linked."""
        return Envelope(type=type_, payload=payload,
                        correlation_id=self.correlation_id, causation_id=self.id,
                        actor=self.actor)
